### PR TITLE
Actionable notifications

### PR DESCRIPTION
**Pull request by alameenshah**
#### High level description
Native notifications can now have actions associated with them. When user clicks on the action button on a notification bubble, `VFSForGit.app` will run a Command in response.
Example: `Mount_Failure` notification can specifiy the `gvfs` command `gvfs mount <path>` as its associated action. Actionable notifications also have an `actionTitle`, which get displayed as text in a Button control on the Notification bubble.

#### More info
- Commands are run using `Terminal` app.
- Terminal is launched using `Applescript`.
- Only `UpgradeAvailable` and `Mount_Failure` notifications are actionable now.
- `VFSForGitNotification` class now confirms to `NSCoding` protocol. `VFSForGitNotification` object holds details of the notification including its associated commands. It is archived using `NSCoding` protocol and sent to the OS notification center. When user clicks the notification bubble, OS Notification Center calls `VFSForGit.app` with the specific notification UI object that user clicked. `VFSForGit.app` then retrieves the archived `VFSForGitNotification` object contained in the notification UI. `VFSForGit.app` then un-archives it, reads the command info contained in it and launches `Terminal` and executes the command.

#### Screenshots
- Upgrade available notification
Screen Shot 2019-06-20 at 2 40 59 PM

- Clicking on `Upgrade` launches the following `Terminal` window - 

Screen Shot 2019-06-20 at 2 41 19 PM

### CODE REVIEW

File: GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/IPC/VFSForGitNotification.m [hunk 1]
```diff
@@ -181,4 +217,60 @@ - (instancetype)initAsMountWithMessage:(NSDictionary *)messageDict
...
+    {
+        id newVersion = messageDict[NewUpgradeVersionKey];
+        if (newVersion && [newVersion isKindOfClass:[NSString class]])
+        {
+            _title = [[NSString stringWithFormat:UpgradeAvailableTitleFormat, newVersion] copy];
+            _message = [UpgradeAvailableMessage copy];
+            _actionTitle = [UpgradeActionTitle copy];
+            _actionable = YES;
+            _gvfsCommand = [UpgradeGVFSCommandFormat copy];
+            return self;
```
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/IPC/VFSForGitNotification.m]: I've removed it from this PR (and couple other places.) In `VFSForGit` multiple returns from a method is a common pattern - so there still are so many other places remaining.

File: GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m [hunk 2]
```diff
@@ -1,42 +1,73 @@
...
+                                              unarchiveObjectWithData:[[notification userInfo]
+                                                                       objectForKey:@"VFSForGitNotification"]];
+    if (vfsNotification == nil)
+    {
+        return;
+    }
+    
+    switch (notification.activationType)
+    {
+        case NSUserNotificationActivationTypeContentsClicked:
```
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m]: For OS upgrades, clicking on the Content triggers launch of Software Update App. In our case we don't have a UI. I'll discuss with the team on the options available.

File: GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m [hunk 3]
```diff
@@ -1,42 +1,73 @@
...
-                              notificationCenter:notificationCenter];
+    return self;
 }
 
-- (instancetype)initWithUserNotification:(NSUserNotification *)userNotification
-                      notificationCenter:(NSUserNotificationCenter *)notificationCenter
+- (void)display:(VFSForGitNotification *) notification
 {
-    if (self = [super init])
+    NSUserNotification *userNotification = [[NSUserNotification alloc] init];
```
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m]: Discussed with @wilbaker and @nickgra . `VFSForGit` right now supports `10.13`. But `UserNotifications` are not available on `10.13`. Couple options were considered.
1. Implement [availability](https://developer.apple.com/documentation/swift/objective-c_and_c_code_customization/marking_api_availability_in_objective-c) check to incorporate new `UserNotifications` - on `10.13` use `NSUserNotification` and on `10.14+` use `UserNotifications`. There will be two separate code paths for `10.13` and `10.14+`.
2. Keep the current implementation. When `VFSForGit` drops support for `10.13` in the near future, replace the old API with the new one. The idea is to not keep separate code paths.

We decided to go with Option 2. In the meanwhile `NSUserNotification` works on `10.13`, `10.14` and `10.15 Beta`.

File: GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m (hunk 2 above)
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m]: Update - discussed with @wilbaker and @nickgra . We decided to ignore Content click for the time being. A new intermediate Window, with upgrade version information plus an `Update` action button, can be added in a separate PR.

File: GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m (hunk 3 above)
[the remark this answers is not in the document]
Comment (HiKumar-MS) [reply GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m]: Thanks for following up. That plan sounds fine to me.

[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m]: > We decided to go with Option 2. In the meanwhile NSUserNotification works on 10.13, 10.14 and 10.15 Beta.

@alameenshah if there's not already a work item for this follow up work, could you file one?

[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/UI/VFSNotificationDisplay.m]: Updated the Notifications [TODO](https://github.com/microsoft/VFSForGit/issues/1262) with a new action item for this

File: GVFS/GVFS.Common/NamedPipes/NamedPipeMessages.cs [hunk 4]
```diff
@@ -235,7 +235,8 @@ public enum Identifier
                 {
                     AutomountStart,
                     MountSuccess,
-                    MountFailure
+                    MountFailure,
+                    UpgradeAvailable
```
Comment (wilbaker) [inline GVFS/GVFS.Common/NamedPipes/NamedPipeMessages.cs]: Is anything sending the `UpgradeAvailable` notifications right now?  I could only find where `MountFailure` is being sent:

https://github.com/microsoft/VFSForGit/blob/27758788bc228371fffd0a6ba968e9ee17f61177/GVFS/GVFS.Service/RepoRegistry.cs#L186

How were you able to test this locally?

File: GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/Info.plist [hunk 5]
```diff
@@ -4,6 +4,8 @@
 <dict>
 	<key>LSApplicationCategoryType</key>
 	<string></string>
+	<key>NSAppleEventsUsageDescription</key>
+	<string>`$(PRODUCT_NAME)` uses `Terminal` app to run `gvfs upgrade` and `gvfs mount` commands.</string>
```
[the remark this answers is not in the document]
Comment (alameenshah) [reply GVFS/GVFS.Notifications/VFSForGit.Mac/StatusMenuItem/Info.plist]: makes sense. [updated](https://github.com/microsoft/VFSForGit/pull/1289/commits/37fda700fde3e1418830df8f847b04942744d5ea).

File: GVFS/GVFS.Common/NamedPipes/NamedPipeMessages.cs (hunk 4 above)
Comment (alameenshah) [reply GVFS/GVFS.Common/NamedPipes/NamedPipeMessages.cs]: `UpgradeAvailable` is not hooked up yet. For testing I use [this branch](https://github.com/alameenshah/VFSForGit/tree/notification_user_actions_notify_verb) in my fork - It has a `notify` verb that can send all supported notifications to the UI.

**Pull request merged**